File: order-model/src/main/scala/ordermodel/usecases/PlaceNewOrderUsecase.scala
package ordermodel.usecases

import ordermodel.domain.Order
import ordermodel.repository.OrderRepository

import scala.beans.BeanProperty

/**
  * @author Bhuwan Upadhyay
  *
  */
class PlaceNewOrderUsecase {

  @BeanProperty
  var orderRepository: OrderRepository = _
  @BeanProperty
  var validateOrderUsecase: ValidateOrderUsecase = _

  def newOrder(newOrder: Order): List[String] = {
    val list = validateOrderUsecase.validate(newOrder)
    if (list.isEmpty)
      orderRepository.add(newOrder)
    list
  }
}

File: order-model/src/main/java/ordermodel/usecases/PlaceNewOrderUsecase.java
package ordermodel.usecases;


import lombok.Setter;
import ordermodel.domain.Order;
import ordermodel.repository.OrderRepository;

import java.util.List;


/**
 * @author Bhuwan Upadhyay
 */
public class PlaceNewOrderUsecase {

    @Setter
    private OrderRepository orderRepository;
    @Setter
    private ValidateOrderUsecase validateOrderUsecase;

    public List<String> newOrder(Order newOrder) {
        List<String> list = validateOrderUsecase.validate(newOrder);
        if (list.isEmpty()) {
            orderRepository.add(newOrder);
        }
        return list;
    }
}
